File: app/javascript/channels/room_channel.js
import consumer from "./consumer"

// Создаем подписку на канал RoomChannel. У объекта есть коллбэки
// connected, disconnected, received, speak.Они действуют как обычные руби - методы.
document.addEventListener('turbo:load', () => {
    let messages

    messages = document.getElementById('messages')

    if (messages) {
        createRoomChannel(messages.dataset.roomId)
    }

    const message = document.getElementById('message_body')

    if (message) {
        messages.scrollTop = messages.scrollHeight

        message.addEventListener('keydown', (event) => {
            if (event.key === 'Enter') {
                if (message.value.length > 0) {
                    room.speak(message.value)
                    message.value = ''
                }

                event.preventDefault()
            }
        })
    }
})

let room

const createRoomChannel = (roomId) => {
    room = consumer.subscriptions.create({
        channel: 'RoomChannel',
        roomId: roomId
    }, {
        connected() {
            console.log('Connected to RoomChannel')
        },

        disconnected() {
            console.log('Disonnected from RoomChannel')
        },

        // получаем сообщение и достаем его из ключа "message"
        received(data) {
            console.log(data)
            messages.innerHTML += data['message']
            messages.scrollTop = messages.scrollHeight
        },

        speak: function(message) {
            return this.perform('speak', {
                message: message
            })
        }
    })
}

// функция "speak" принимает на вход сообщение. Шлем его с пом-ю этого метода.
// Этот метод вызывает действие "speak" на серверном канале
// метод "perform" дергает соответствующий метод канала на сервере

File: app/channels/room_channel.rb
# frozen_string_literal: true

class RoomChannel < ApplicationCable::Channel
  # вызывается, когда клиент по этому каналу подсоединился к серверу
  def subscribed
    # регистратор ("подписан на канал RoomChannel")
    logger.info "Subscribed to RoomChannel, roomId: #{params[:roomId]}"

    @room = Room.find(params[:roomId])

    # передать в JS в соответствующую комнату
    # сообщения сервер разводит по разным комнатам
    stream_from "room_channel_#{@room.id}"
  end

  # в противном случае
  def unsubscribed
    logger.info 'Unsubscribed from RoomChannel'
  end

  def speak(data)
    # получем данные от пользователя, выводим их в лог нашего сервера
    logger.info "Speak: #{data.inspect}"

    # что мы шлём
    MessageService.new(
      body: data['message'],
      room: @room,
      user: current_user
    ).perform
  end
end
